caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h

```diff
@@ -23,28 +23,24 @@
 #include <assert.h>
 #include <stdio.h>
 
-#define DEBUG_PRINT(...) ;
+#define DEBUG_PRINT(...)
 
 namespace nom {
 
-template <typename T, typename U = T>
+template <typename T, typename U>
 class Graph;
 
-template <typename T, typename U = T>
+template <typename T, typename U>
+class GraphBase;
+
+template <typename T, typename U>
 class Node;
 
-template <typename T, typename U = T>
-class Edge : public StorageType<U> {
+template <typename T, typename U>
+class EdgeBase {
```

Comment:
have you tried applying this "base" paradigm to the StorageType class itself?  might clean things up here.  The duplication of of constructors is a bit messy

Reply:
I am not sure (conceptually) how this would work, though. 

To brain storm: 
Say, C++ figured out all the types, and we have actual template instances: We'd have an empty StorageType class:
class EmptyStorageType {};
Now, at the higher layer, we'd have the Edge class. Let's say, for the void case, it inherits from this:
class EdgeEmpty : public EmptyStorageType, public EdgeBase {...}

Now, for constructors for EdgeEmpty:
case 1) There is a three-arg constructor. In this case, the third argument is going to have the void type, and this is not allowed.
case 2) There is no three-arg constructor: In this case, the non-Empty Edge class needs to have a three-arg constructor. Therefore, I still need to have two classes. To avoid duplicating code, I'd still have an EdgeBase, and we come to the same EdgeBase + 2 Edge version, except now we have an extra empty StorageType class that doesn't contain anything.

Do you see anything that I'm missing?

Reply:
I was thinking `Edge` can remain unchanged and inherit from `StorageType<void>` which is a specialized class that doesn't do anything

Reply:
just had a thought that you can use a parameter pack to get it working.  I'm not entirely sure how the compiler chooses a constructor, to be quite honest...

```
diff --git a/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h b/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
index b9b338f37..f513887ca 100644
--- a/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
+++ b/caffe2/core/nomnigraph/include/nomnigraph/Graph/Graph.h
@@ -33,7 +33,7 @@ class Graph;
 template <typename T, typename U = T>
 class Node;

-template <typename T, typename U = T>
+template <typename T, typename U = void>
 class Edge : public StorageType<U> {
  public:
   using NodeRef = typename Graph<T, U>::NodeRef;
@@ -43,8 +43,9 @@ class Edge : public StorageType<U> {
     DEBUG_PRINT("Creating instance of Edge: %p\n", this);
   }

-  Edge(NodeRef tail, NodeRef head, U&& data)
-      : StorageType<U>(std::move(data)), Tail(tail), Head(head) {
+  template <typename... Args>
+  Edge(NodeRef tail, NodeRef head, Args... args)
+      : StorageType<U>(std::forward<Args>(args)...), Tail(tail), Head(head) {
     DEBUG_PRINT("Creating instance of Edge: %p\n", this);
   }

diff --git a/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h b/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
index 0f0ba9545..095e30143 100644
--- a/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
+++ b/caffe2/core/nomnigraph/include/nomnigraph/Support/Common.h
@@ -35,6 +35,12 @@ class StorageType {
   T Data;
 };

+template <>
+class StorageType<void> {
+public:
+  StorageType() {}
+};
+
 /// \brief This class enables a listener pattern.
 /// It is to be used with a "curious recursive pattern"
 /// i.e. Derived : public Notifier<Derived> {}
```